libp2p/connection.nim

```diff
@@ -119,30 +127,37 @@ method closed*(s: Connection): bool =
   result = s.stream.closed
 
 method close*(s: Connection) {.async, gcsafe.} =
-  trace "about to close connection", closed = s.closed, conn = $s
+  try:
+    if not s.isClosed:
+      s.isClosed = true
 
-  if not s.isClosed:
-    s.isClosed = true
-    inc getConnectionTracker().closed
+      trace "about to close connection", closed = s.closed,
+                                        conn = $s,
+                                        oid = s.oid
 
-    if not isNil(s.stream) and not s.stream.closed:
-      trace "closing child stream", closed = s.closed, conn = $s
-      try:
-        await s.stream.close()
-      except CancelledError as exc:
-        raise exc
-      except CatchableError as exc:
-        debug "Error while closing child stream", err = exc.msg
 
-    s.closeEvent.fire()
-
-    trace "waiting readloops", count=s.readLoops.len, conn = $s
-    let loopFuts = await allFinished(s.readLoops)
-    checkFutures(loopFuts)
-    s.readLoops = @[]
-
-    trace "connection closed", closed = s.closed, conn = $s
-    libp2p_open_connection.dec()
+      if not isNil(s.stream) and not s.stream.closed:
+        trace "closing child stream", closed = s.closed,
+                                      conn = $s,
+                                      oid = s.stream.oid
+        await s.stream.close()
+        # s.stream = nil
+
+      s.closeEvent.fire()
+      trace "waiting readloops", count=s.readLoops.len,
+                                 conn = $s,
+                                 oid = s.oid
+      await all(s.readLoops)
```

Comment:
Aren't you supposed to check the futures in `s.readLoops` for errors, after waiting for their completion?

Reply:
Well `all` should throw and that's the reason to use it, in that case the wrapping except would catch it, but the funny thing is that `readLoops` isn't even used anywhere, this was one of @sinkingsugar changes and its purpose escapes me now. Good catch, thanks @stefantalpalaru .

Reply:
Oh, I remembered why it's not needed anymore, I'm not using the internal close event anymore to propagate upstream closes, so that is gone now.

Reply:
> Well `all` should throw and that's the reason to use it

Futures don't throw exceptions, but encapsulate them in an `error` field: https://github.com/status-im/nim-chronos/blob/bce0f878d16bea14395dc8944ff8dbdeffe33496/chronos/asyncfutures2.nim#L34

The `all()` procedure does the same, by setting the same `error` field in the new Future it returns: https://github.com/status-im/nim-chronos/blob/bce0f878d16bea14395dc8944ff8dbdeffe33496/chronos/asyncfutures2.nim#L544

Reply:
read the asyncmacro2.nim `await`, for your convenience here is the code:

```nim
template await*[T](f: Future[T]): auto =
  when declared(chronosInternalRetFuture):
    when not declaredInScope(chronosInternalTmpFuture):
      var chronosInternalTmpFuture {.inject.}: FutureBase
    chronosInternalTmpFuture = f
    chronosInternalRetFuture.child = chronosInternalTmpFuture
    yield chronosInternalTmpFuture
    chronosInternalTmpFuture.internalCheckComplete()
    chronosInternalRetFuture.child = nil
    cast[type(f)](chronosInternalTmpFuture).internalRead()
  else:
    unsupported "await is only available within {.async.}"
```

What will this do?

Reply:
> What will this do?

It will not raise any exception past the async boundary, if that's what you're thinking, because `await` can only be used in async procs which convert all exceptions in object fields.

Proof:

```nim
import chronos

proc foo() {.async.} =
  echo "in foo()"
  if true: # work around a {.noReturn.} pragma
    raise newException(CatchableError, "foo")

proc bar() {.async.} =
  echo "in bar()"
  await foo()

let fut = bar()
while not fut.finished():
  poll()
```

This will not raise anything. The `await()` in `bar()` only transferred the exception transformed into a `Future[void].error` in `foo()` to another `Future[void].error` returned by `bar()`.

Add `echo fut.error.msg` at the end of this example to see for yourself.

Reply:
you're not reading the future, it will not throw, in other words you're using it incorrectly and expecting the lib to behave consistently.

This is how you do it:
```nim
import chronos

proc main() {.async.} =
  proc foo() {.async.} =
    echo "in foo()"
    if true: # work around a {.noReturn.} pragma
      raise newException(CatchableError, "foo")

  proc bar() {.async.} =
    echo "in bar()"
    await foo()

  await bar()

var mainFut = main()
while not mainFut.finished():
  poll()

mainFut.read()
```

Reply:
But that's the reason you don't do it manually, and use something like `waitFor`, if you aren't using the proper API then expect things to break. Not a knock you your example, it just _IS_ improper use of the api.

Reply:
> you're not reading the future, it will not throw

Excellent! Now show me where you read the Future in this try block: https://github.com/status-im/nim-libp2p/pull/187/files#diff-43d36f9fd0a52d64fd707e3802cae10eR89-R93

Reply:
> it just IS improper use of the api

No, it's not. Both `poll()` and `runForever()` are part of the Chronos public API. None of them call `Future.read()` for you. None of them can be replaced by `waitFor()` in all legitimate scenarios.

Reply:
> No, it's not. Both poll() and runForever() are part of the Chronos public API. None of them call Future.read() for you. None of them can be replaced by waitFor() in all legitimate scenarios.

Yes, and as  you showed there are legitimate use cases for them - it's just not everywhere. Is it a gotcha, probably, but it can be partially addressed with proper documentation. 

Is it a gotcha now, definitely yes because you have to read the code, which is cumbersome and makes it easy to miss things, good example is my misreading/misinterpretation of `allFutures`.

But hey,  we are reinventing the wheel pretty much everywhere and this is definitely not in my list of major issue, but I agree that it needs to be addressed.